### ISSUE TITLE
hcloud server ssh wait

### ISSUE THREAD

**Issue by ktomk** [hetznercloud/cli#273]
It would be great if there are some timeout / setting parameters to open an ssh session more robust when the virtual machine has been just powered on.

Currently for scripting it's kind of akward, like in the following example:

    $ hcloud server poweron $HCLOUD_SERVER \
        && hcloud server ssh $HCLOUD_SERVER 'echo "ssh connected"' \
        || sleep 3 \
        ; hcloud server ssh $HCLOUD_SERVER 'echo "ssh connected"' \
        || sleep 2 \
        ; hcloud server ssh $HCLOUD_SERVER 'echo "ssh connected"' \
        || sleep 1 \
        ; hcloud server ssh $HCLOUD_SERVER \
            'apt-get update && apt-get -u upgrade -y '\
            '&& npm install -g npm; shutdown -h now; sleep 3'

Apart from that the example might be convoluted/theoretic for other reasons (why poweron/off), I hope it makes the scenario/request visible what I mean.

**Comment by LKaemmerling** [comment]
Hey @ktomk,

i think this is out of the scope for the cli. We can never see if the server is "ready" to accept ssh connections (or if he is doing some other stuff). I would recommend using cloud init instead if doing the thing you tried. 

(You could install your npm as a sample with cloud init), but actually this won't solve the "ssh/server is not ready issue". 

Implementing a retrying into the ssh command won't fit i think. @fhofherr what do you think?

**Comment by ktomk** [comment]
Hey @LKaemmerling,

thanks for your open feedback.

Yes, it is technically not possible to say if the server is ready to accept SSH connections. Something could be inherently wrong so the server is not available (e.g. error), impossible to distinguish from it *might* become available.

The base concept was more to specify the number of retries to connect to ssh with some timing options between retries. Something like the [`retry` command on Travis-CI][travis_retry] for example. But it would not work for the full fledged SSH command in my example as it exists with non-zero status when successful. Just saying, I'm aware it's a pretty specific example to that regard and should if done be really simplified - dead simple.

Honestly, I also was a bit lazy. Somewhat later I eventually discovered that `server ssh` offers me all the open SSH options and I should - but didn't so far - look into there if it's not something SSH already offers. Will update for that (I hope I have not seen that wrong).

Your suggestion of Cloudinit is also something I was dipping into for Hetzner Cloud, merely for default setup/bootstrap, but it may blend in well in my use-case.

One reason I would say that why retrying in the `server ssh` command would fit is that I don't need another abstraction layer. Especially for the SSH command, it is a gate-opener for remote-commands on the server and would help to make more use out of the `hcloud` utility as these commands send via SSH would become more stable, I'm not in the need to add another layer of abstraction around for that, like Ansible and beyond. Currently I'm learning how effective the `hcloud` utility is already, have been doing more Ansible stuff earlier with Hetzner Cloud which all in all works well, no complains on the hcloud module there, it's just that Ansible felt more and more standing in the way then being helpful. And SSH is the key connector in Ansible (also was fighting with the dynamic inventory, it's not easy to combine with other inventories which I think is a limitation of Ansible itself as well, less a problem of the Hetzner Cloud inventory plugin - anyway, off-topic).

So for more lower-level tooling and automation, would be really nice to not care about early connection failures and to not have it necessary to do some bulky sleep 3 on good luck (which then might become 4, 7 or 10 even over time). Brings me back to the point, that it would be best if I can express it with SSH switches / options already. Looking into that now.

References:

- [`travis_rety` shell function][travis_retry] (by Travis; on Github)
- [Connect to an SSH server until it works](https://musigma.blog/2016/12/03/ssh-auto-retry.html) (by Musigma; Dec, 2016)

[travis_retry]: https://github.com/travis-ci/travis-build/blob/73f74a94957f73eb54dc821f80c0c85ad8f8aab7/lib/travis/build/script/templates/header.sh#L168-L187

**Comment by fhofherr** [comment]
Hi @ktomk,

thank you for your request. As @LKaemmerling already said we do not consider this as in-scope for the `hcloud` CLI tool. I guess it is easier for the users to just implement a small helper script or shell function that keeps retrying to establish the connection until it is successful. Something similar to this:

```bash
#!/usr/bin/env bash

function retry_ssh() {
    local server_name ec

    if (( $# == 0)); then
        echo "Usage:"
        echo "    $0 ssh <server-name> [command]"
    fi
    server_name="$1"
    shift

    for (( i=0; i < 3; i++ )); do
        hcloud server ssh "$server_name" "$@"
        ec=$?

        if (( ec == 0 )); then
            return 0
        fi
        sleep "$i"
    done
    return $ec
}

retry_ssh "$@"
```

SSH should always execute with the status code of the last command executed. Therefore you would need to ensure that whatever you do during retry would always return with a 0 exit code:

```bash
if ! retry.sh "$SERVER_NAME" echo "I'm there"; then
    echo "$SERVER_NAME startup failed"
   exit 1
fi

hcloud server ssh <commands that might fail>
```

**Comment by ktomk** [comment]
Hey @fhofherr. thanks a lot for your feedback. And good news: not even a helper script is needed as it turns out: the retry feature is already built-in, my day before yesterdays idea shows working good already.

Found the SSH options and it is a breeze. The example from above with automatic retries and connection timeout settings from SSH

```
$ hcloud server poweron $HCLOUD_SERVER \
    && hcloud server ssh $HCLOUD_SERVER -o ControlMaster=no -o ConnectTimeout=3 -o ConnectionAttempts=9 \
        'export DEBIAN_FRONTEND=noninteractive ;'\
        'apt-get update && apt-get -u --with-new-pkgs upgrade -y '\
        '&& npm install -g npm; shutdown -h now;' \
    ; test $? -eq 255
```

The magic is in the `ConnectionAttempts` SSH option:

```
$ hcloud server ssh $HCLOUD_SERVER  -o ControlMaster=no -o ConnectTimeout=3 -o ConnectionAttempts=9
                                   ^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^
```

(the following quotes are all from `man ssh_config`)

> **ConnectionAttempts**
>
>   Specifies the number of tries (one per second) to make before exiting.  The argument must be an integer.  This may be useful in scripts if the connection sometimes fails. The default is 1.

The number of retries. As this means one second per retry, a relatively high number can be used as if connection is established earlier, it does not take long.

> **ConnectTimeout**
> 
> Specifies the timeout (in seconds) used when connecting to the SSH server, instead of using the default system TCP timeout.
>
> This value is used only when the target is down or really unreachable, not when it refuses the connection.

If the server just started, it is not yet reachable. Using a small number keeps ongoing re-tries fast. One second to time-out, plus one second wait time before the next connection attempt.

> **ControlMaster**
>
> Enables the sharing of multiple sessions over a single network connection. When set to yes, ssh(1) will listen for connections on a control socket specified using the ControlPath argument.  Additional sessions can connect to this socket using the same ControlPath with ControlMaster set to no (the default). [...]

Set to the default value just in case it was activated in the inherited configuration. Multiplexing can make connections hang. This is a precaution of a non-default default SSH configuration (e.g. SSH Multiplexing enabled by default).

In my scenario I ended up at 9 connection attempts with a 3 second timeout value.

One little detail I ran over with this configuration, I had to mind the OpenSSH version: 7.6 introduced a flaw rendering ConnectTimeout useless, as it was only for the first attempt, any other attempt timed out in ~130 seconds (system default?). With nine connection attempts this can become a good share of an hour. The bug ([Bug 2918 - ssh ConnectTimeout is obeyed only on 1st ConnectionAttempt](https://bugzilla.mindrot.org/show_bug.cgi?id=2918)) was fixed in 8.0.

For learning about timings, `-v` was helpful as it shows the retries.

So this is now my working variant and a very good resolution.

Thanks again for all your feedback you gave especially it is not at all an hcloud issue (neither fix nor feature).
